src/mlpack/methods/decision_tree/decision_tree.hpp

```diff
@@ -496,6 +505,8 @@ class DecisionTree :
   std::vector<DecisionTree*> children;
   //! The dimension this node splits on.
   size_t splitDimension;
+  //! The best gain on the particular node.
+  double nodeGain;
```

Comment:
I hate to raise this comment again, because it makes the implementation of this much more difficult, but the `DecisionTree` class is carefully constructed to be as small as possible.  This is very important for prediction with decision trees as well as serialization, as there is one of these objects for every single node in the tree.  Additionally storing 8 bytes in each node for `nodeGain` adds up quickly---it's not unreasonable for a tree to have 1M nodes in it, and that in turn means 8MB of size gain, even for trees that do not need `nodeGain`.

In any case, isn't this functionality already handled by `minimumGainSplit`?

Or if this needs to be a separate step for XGBoost, where you train the tree, then find a threshold, then need to prune the tree, we actually do have a way to do this already: via the `NumericAuxiliarySplitInfo` and `CategoricalAuxiliarySplitInfo` classes.  See `NumericSplitType` in [the documentation](https://www.mlpack.org/doc/user/methods/decision_tree.html#advanced-functionality-template-parameters)).

So, for XGBoost splits, you could use a split type that stores `nodeGain` in its `AuxiliarySplitInfo` internal class.  This does make the `DecisionTree` class larger (again by 8 bytes for each node), but it is only when XGBoost splitting strategies are used.

Then, you could even recover the 8 bytes per node by having `Prune()` operate as a standalone method that takes in an unpruned XGBoost tree (whose split types have an `AuxiliarySplitInfo` that holds `nodeGain`), and emits a new XGBoost tree with a *different* split type whose `AuxiliarySplitInfo` is empty---as the node gain is no longer needed.

I know that what I have written here looks like an absolutely crazy complicated design (and... well, to some extent, it is!), but I don't really see another easy way to avoid that extra size, which does matter (and this type of situation is why `AuxiliarySplitInfo` was designed the way it is).

Reply:
Oh yeah, good point.

I mainly want to implement Pruning because it can become more complex, including more advanced stuff like Stochastic Pruning, Regularized Pruning etc. Right now I've just done a basic implementation to get started.

I'll definitely check out your suggestion about `AuxiliarySplitInfo`.
Thanks!